Neos.Neos/Classes/Domain/Workspace/Workspace.php

```diff
@@ -0,0 +1,445 @@
+            if ($ancestorNodeTypeName->equals(NodeTypeNameFactory::forSite())) {
+                return true;
+            }
+        }
+
+        $subgraph = $this->contentRepository->getContentGraph()->getSubgraph(
+            $this->currentContentStreamId,
+            $change->originDimensionSpacePoint->toDimensionSpacePoint(),
+            VisibilityConstraints::withoutRestrictions()
+        );
+
+        // A Change is publishable if the respective node (or the respective
+        // removal attachment point) has a closest ancestor that matches our
+        // current ancestor scope (Document/Site)
+        $actualAncestorNode = $subgraph->findClosestNode(
+            $change->removalAttachmentPoint ?? $change->nodeAggregateId,
+            FindClosestNodeFilter::create(nodeTypes: $ancestorNodeTypeName->value)
+        );
+
+        return $actualAncestorNode?->nodeAggregateId->equals($ancestorId) ?? false;
+    }
+
+    /**
+     * Before the introduction of the WorkspacePublisher, the UI only ever
+     * referenced the closest document node as a removal attachment point.
+     *
+     * Removed document nodes therefore were referencing themselves.
+     *
+     * In order to enable publish/discard of removed documents, the removal
+     * attachment point of a document MUST refer to an ancestor. The UI now
+     * references the site node in those cases.
+     *
+     * Workspaces that were created before this change was introduced may
+     * contain removed documents, for which the site node can longer be
+     * located, because we have no reference to their respective site.
+     *
+     * Every document node that matches that description will be published
+     * or discarded by WorkspacePublisher::publishSite, regardless of what
+     * the current site is.
+     *
+     * @deprecated remove once we are sure this check is no longer needed due to
+     * * the UI sending proper commands
+     * * the ChangeFinder being refactored / rewritten
+     * (whatever happens first)
+     */
+    private function isChangeWithSelfReferencingRemovalAttachmentPoint(Change $change): bool
+    {
+        return $change->removalAttachmentPoint?->equals($change->nodeAggregateId) ?? false;
+    }
+
+    private function updateCurrentState(): void
+    {
+        /** The workspace projection should have been marked stale via @see WithMarkStaleInterface in the meantime */
+        $contentRepositoryWorkspace = $this->contentRepository->getWorkspaceFinder()
+            ->findOneByName($this->name);
+        if (!$contentRepositoryWorkspace) {
+            throw new WorkspaceDoesNotExist('Cannot update state of non-existent workspace ' . $this->name->value, 1711704397);
+        }
+
+        $this->currentContentStreamId = $contentRepositoryWorkspace->currentContentStreamId;
+        $this->currentStatus = $contentRepositoryWorkspace->status;
+        $this->currentBaseWorkspaceName = $contentRepositoryWorkspace->baseWorkspaceName;
+    }
```

Comment:
i guess we get such mutations for free when using entity like patterns :O

For the `subgraph` we decided that it will use the `currentContentStreamId` of its build time (`getSubgraph`) instead of refreshing itself.
Does it make sense to use a different strategy here? I guess so because otherwise the workspace thingy can only be used to call one method and would need to be discared otherwise ...

Reply:
We currently operate in most parts in a immutable fashion. The upcoming `getSubgraph` api will behave the following

```php
// current mapping workspace name to content stream id
// 'user-admin' => 123

$subgraph = $contentGraph->getSubgraph('user-admin');
assert($subgraph === $contentGraph->getSubgraph('user-admin')); // result is cached until `handle`.
assert($subgraph->getIdentity()->getContentStreamId() === 123);
$subgraph->fetchSomething();  // operates on 123

$cr->handle(RebaseWorkspace('user-admin'));
// handle flushes via markStale the cached $subgraph instances retrieved via `getSubgraph`
// updated workspace name to content stream id
// 'user-admin' => 456

// the old subgraph still operates on 123
assert($subgraph->getIdentity()->getContentStreamId() === 123);
$subgraph->fetchSomething();  // operates still on 123

$newSubgraph = $contentGraph->getSubgraph('user-admin');
assert($subgraph->getIdentity()->getContentStreamId() === 456);
$subgraph->fetchSomething();  // operates still on 456
```

I was referring to when can the user expect the object to behave "intelligent" / mutating and when not.

Reply:
I'd say the high level API should behave "intelligently", i.e. after rebase, getCurrentState on the same object should return UP-TO-DATE

Reply:
Exactly i agree as we are going with this entity like mutable road, but wanted to point out this difference so no one is surprised ;)

Reply:
Made them singletons and added this line

> This model is mutable and will for example update itself after publishing, or changing the base workspace.
Mutations in the content repository that are not triggered by this model (by using the low level API) will not be reflected.


see  https://github.com/neos/neos-development-collection/pull/4974

Reply:
Nice, merged :)